shared/controlflow/codeql/controlflow/BasicBlock.qll

```diff
@@ -0,0 +1,370 @@
+     */
+    predicate strictlyDominates(BasicBlock bb) { bbIDominates+(this, bb) }
+
+    /**
+     * Holds if this basic block dominates basic block `bb`.
+     *
+     * That is, all paths reaching `bb` from the entry point basic block must
+     * go through this basic block.
+     */
+    predicate dominates(BasicBlock bb) {
+      bb = this or
+      this.strictlyDominates(bb)
+    }
+
+    /**
+     * Holds if `df` is in the dominance frontier of this basic block. That is,
+     * this basic block dominates a predecessor of `df`, but does not dominate
+     * `df` itself. I.e., it is equivaluent to:
+     * ```
+     * this.dominates(df.getAPredecessor()) and not this.strictlyDominates(df)
+     * ```
+     */
+    predicate inDominanceFrontier(BasicBlock df) {
+      // Algorithm from Cooper et al., "A Simple, Fast Dominance Algorithm" (Figure 5),
+      // who in turn attribute it to Ferrante et al., "The program dependence graph and
+      // its use in optimization".
+      this = df.getAPredecessor() and not bbIDominates(this, df)
+      or
+      exists(BasicBlock prev | prev.inDominanceFrontier(df) |
+        bbIDominates(this, prev) and
+        not bbIDominates(this, df)
+      )
+    }
+
+    /**
+     * Gets the basic block that immediately dominates this basic block, if any.
+     *
+     * That is, the result is the unique basic block satisfying:
+     * 1. The result strictly dominates this basic block.
+     * 2. There exists no other basic block that is strictly dominated by the
+     *    result and which strictly dominates this basic block.
+     *
+     * All basic blocks, except entry basic blocks, have a unique immediate
+     * dominator.
+     */
+    BasicBlock getImmediateDominator() { bbIDominates(result, this) }
+
+    /**
+     * Holds if basic block `succ` is immediately controlled by this basic
+     * block with successor type `s`.
+     *
+     * That is, `succ` is an immediate successor of this block, and `succ` can
+     * only be reached from the entry block by going via the `s` edge out of
+     * this basic block.
+     */
+    pragma[nomagic]
+    predicate immediatelyControls(BasicBlock succ, SuccessorType s) {
+      succ = this.getASuccessor(s) and
+      bbIDominates(this, succ) and
+      // The above is not sufficient to ensure that `succ` can only be reached
+      // through `s`. To see why, consider this example corresponding to an
+      // `if` expression without an `else` block:
+      // ```
+      // ... --> cond --[true]--> ... --> if expr
+      //           \                      /
+      //            ----[false]-----------
+      // ```
+      // The basic block for `cond` immediately dominates the directly
+      // succeeding basic block for the `if` expression. But the `if`
+      // expression is not immediately controlled by the `cond` basic block and
+      // the `false` edge since it is also possible to reach the `if`
+      // expression via the `true` edge.
```

Comment:
Maybe just replace `if expr` with a name like e.g. `A` in the ascii graph to identify the subsequent basic block?
```suggestion
      // `if` statement without an `else` block:
      // ```
      // ... --> cond --[true]--> ... --> A
      //           \                      /
      //            ----[false]-----------
      // ```
      // The basic block for `cond` immediately dominates the basic block `A`
      // that follows the `if` statement. But `A`
      // is not controlled by the `cond` basic block and
      // the `false` edge since it is also possible to reach `A`
      // via the `true` edge.
```